File: kubi-api/src/routes/similar.ts
import { Hono } from "hono";
import { z } from "zod";
import { callKubiService } from "../lib/client";

const similarSchema = z.object({
  url: z.string().url(),
  query: z.string().optional(),
  type: z.enum(["instant", "fast", "auto", "deep-lite", "deep", "research"]).default("fast"),
  numResults: z.number().int().min(1).max(100).default(10),
  includeDomains: z.array(z.string()).optional(),
  excludeDomains: z.array(z.string()).optional(),
});

export const similarRoutes = new Hono();

similarRoutes.post("/", async (c) => {
  const raw = await c.req.json();
  const parsed = similarSchema.safeParse(raw);
  if (!parsed.success) return c.json({ error: "Validation error", details: parsed.error.flatten() }, 400);

  const d = parsed.data;
  try {
    const result = await callKubiService({ path: "/ai/similar", body: {
      url: d.url, query: d.query, type: d.type, num_results: d.numResults,
      include_domains: d.includeDomains || [], exclude_domains: d.excludeDomains || [],
    }});
    return c.json(result);
  } catch (err: any) {
    return c.json({ error: "Similar search failed", message: err.message }, 500);
  }
});

File: kubi-core/src/routes/similar.rs
use axum::{extract::State, Json};
use serde_json::json;
use crate::AppState;

pub async fn similar(State(state): State<AppState>, Json(body): Json<serde_json::Value>) -> Json<serde_json::Value> {
    let url = format!("{}/ai/similar", state.kubi_ai_url);
    match state.http_client.post(&url).json(&body).send().await {
        Ok(resp) => match resp.json::<serde_json::Value>().await {
            Ok(result) => Json(result),
            Err(_) => Json(json!({"error": "Failed to parse response"})),
        },
        Err(e) => Json(json!({"error": "Similar failed", "message": e.to_string()})),
    }
}
